nixos/modules/misc/documentation.nix

```diff
@@ -1,30 +1,54 @@
-{ config, lib, pkgs, extendModules, noUserModules, ... }:
+{ config, lib, options, pkgs, baseModules, extraModules, modules, modulesPath, utils, ... }:
 
 with lib;
 
 let
 
   cfg = config.documentation;
+  allOpts = options;
 
   /* Modules for which to show options even when not imported. */
   extraDocModules = [ ../virtualisation/qemu-vm.nix ];
 
-  /* For the purpose of generating docs, evaluate options with each derivation
-    in `pkgs` (recursively) replaced by a fake with path "\${pkgs.attribute.path}".
-    It isn't perfect, but it seems to cover a vast majority of use cases.
-    Caveat: even if the package is reached by a different means,
-    the path above will be shown and not e.g. `${config.services.foo.package}`. */
+  canCacheDocs = m:
+    let
+      f = import m;
+      instance = f (mapAttrs (n: _: abort "evaluating ${n} for `meta` failed") (functionArgs f));
+    in
+      cfg.nixos.splitOptionDocBuild
+        && builtins.isPath m
+        && isFunction f
+        && instance ? options
+        && instance.meta.buildDocsInSandbox or true;
+
+  docModules =
+    let
+      p = partition canCacheDocs (baseModules ++ extraDocModules);
+    in
+      {
+        lazy = p.right;
+        eager = p.wrong ++ optionals cfg.nixos.includeAllModules (extraModules ++ modules);
+      };
+
   manual = import ../../doc/manual rec {
     inherit pkgs config;
     version = config.system.nixos.release;
     revision = "release-${version}";
     extraSources = cfg.nixos.extraModuleSources;
     options =
       let
-        extendNixOS = if cfg.nixos.includeAllModules then extendModules else noUserModules.extendModules;
-        scrubbedEval = extendNixOS {
-          modules = extraDocModules;
-          specialArgs.pkgs = scrubDerivations "pkgs" pkgs;
+        scrubbedEval = evalModules {
+          modules = [ {
+            nixpkgs.localSystem = config.nixpkgs.localSystem;
```

Comment:
I don't think you should need this for docs.

Duplicating the `evalModules` logic is not great for maintainability. Would it be feasible to use `extendModules` when module filtering is supported by the module system itself?

Reply:
you're right, that isn't needed here. wonder where that even came from now, probably from attempts to put `nixpkgs.nix` into the sandbox (which failed anyway) :thinking:

Reply:
You could do essentially `extendModules { modules = [ { config._module.args.check = lib.mkForce false; } ]; }` for `check`.
The only challenge is removing some modules.

Reply:
"some" is a bit of an understatement, there's over 1400 of them at time of writing :smile: filtering loaded modules like that would be possible if that facility existied, but evaluating a few user modules from scratch is probably faster. :thinking: